README

Dpmaster, a master server for DarkPlaces and Quake III Arena
------------------------------------------------------------

* INTRODUCTION:

Dpmaster is a master server written from scratch for LordHavoc's great Quake 1
engine: DarkPlaces. The master protocol used by DarkPlaces and dpmaster is
heavily based on the Quake III Arena master server protocol, that's why
dpmaster supports DP servers and Q3A servers at the same time.

Although dpmaster is being developed primarily on a Linux machine (i386), it
should compile and run at least on any operating system from the Win32 or UNIX
family. But be aware that some options are only available on UNIXes, including
all security-related options. See the SECURITY section below for more
informations.

The source code of dpmaster is available under the GNU General Public License.


* SYNTAX & OPTIONS:

The syntax of the command line is the classic: "dpmaster [options]". Running
"dpmaster -h" will print the available options for your version.

For a simple usage, the only option you may want to know is '-D'. This is a
UNIX-only option and it makes dpmaster running as a system daemon, in the
background.


* SECURITY:

First, you shouldn't be afraid to run dpmaster on your machine: at the time I
wrote those lines, no security flaw has ever been found in any version of
dpmaster. It has always been developed with security in mind and will always
be.

Also, dpmaster needs very few things to run correctly. A little bit of memory,
a few CPU cycles from time to time and a network port are its only
requirements. So feel free to restrict its privileges as much as you can.

The UNIX/Linux version of dpmaster has even a builtin security mechanism that
triggers when it is run with super-user (root) privileges. Basically, the
process locks (chroots) itself in the empty directory "/var/empty/" and drops
its privileges in favor of those of user "nobody" (notorious for having almost
no privileges at all  :)  Note that this path and this user name are
customizable via the '-j' and '-u' command line options.

Finally, you will probably be glad to know that dpmaster just can't leak
memory, by design.


* CHANGES:

    - version 1.1:
        A lot of optimizations and tweakings
        Verbose option added (-v)
        Hash size option added (-H)
        Daemonization option added on UNIXes (-D)
        Chrooting and privileges dropping when running as root added on UNIXes
        MinGW cross-compilation support added

    - version 1.01:
        A major bug was fixed. Most of the servers weren't sent to the clients

    - version 1.0 :
        First publicly available version


* CONTACTS & LINKS:

You can get more informations and the latest versions of DarkPlaces and
dpmaster on the DarkPlaces home page: http://icculus.org/twilight/darkplaces/


--
Mathieu Olivier  <elric@planetblood.com>


changelog.txt

                         Dpmaster, an open master server
                         -------------------------------

                                Version history
                                ---------------


* Version 2.2-dev:
  - Flood protection against abusive clients, by Timothee Besset
  - New system for managing game properties (see GAME PROPERTIES in manual.txt)
  - Support for RtCW and WoET, using the game properties
  - Shutdown heartbeats and unknown heartbeats are now ignored
  - The chroot jail was preventing daemonization (fixed thanks to LordHavoc)
  - The game type was incorrect when printing the server list in the log
  - Less debug output when creating a getserversResponse in verbose mode

* Version 2.1:
  - A gametype value can now be any string, not just a number

* Version 2.0:
  - Gametype filter support in the server list queries (see techinfo.txt)
  - New option "--game-policy" to filter games (see GAME POLICY in manual.txt)
  - IPv6 support, including 2 new messages types (see techinfo.txt)
  - Logging support (see LOGGING in manual.txt)
  - Only the last packet of a getservers response gets an EOT mark now
  - The default number of servers is now 4096
  - Improved listening interface option (see LISTENING INTERFACES in manual.txt)
  - Long format for all command line options (see SYNTAX & OPTIONS in
    readme.txt)
  - The server lists are now sent in a semi-random order, for fairness
  - The new hash function supports up to 16-bit hashes
  - The default hash size has been increased to 10 bits
  - 0 is no longer an invalid hash size
  - New option "--allow-loopback", for debugging purposes only!
  - New option "--hash-ports", for debugging purposes only!
  - Various updates and improvements in the documentation
  - No warning is printed anymore if a server changes its game name
  - No longer tolerates several mapping declarations for the same address
  - A lot of minor changes and fixes in the code
  - The test suite now requires the Socket6 Perl module to run

* Version 1.7:
  - There's now a maximum number of servers per IP address (default: 32)
  - New option to set the maximum number of servers per IP address (-N)
  - The maximum number of servers recorded by default is now 1024
  - The default hash size has been increased from 6 bits to 8 bits
  - A few Perl scripts have been added to provide basic automated testing
  - A rare bug where a server was occasionally skipped was fixed
  - The compilation with MS Visual Studio 2005 is fixed
  - Protocol numbers less than or equal to 0 are now handled correctly
  - Servers can no longer keep their slots without sending infoResponses
  - Games having a name starting with a number are now handled correctly
  - A few minor memory leaks were removed in the address mapping init code
  - Additional checks of the command line options and the messages syntax
  - The requirement of a "clients" value in infoResponses is now enforced
  - The "infoResponse" description in techinfo.txt has been corrected
  - The "heartbeat" description in techinfo.txt has been corrected
  - The time is now printed to the console each time a packet is received
  - Made it clear in the doc that any game can be supported out of the box

* Version 1.6:
  - Several getserversResponse may now be sent for a single getservers
  - A getserversResponse packet can no longer exceed 1400 bytes
  - The maximum number of servers recorded by default has doubled (now 256)
  - The default hash size has been increased from 5 bits to 6 bits
  - Several updates and corrections in the documentation

* Version 1.5.1:
  - Compilation on FreeBSD was fixed
  - A couple of minor changes in "COMPILING DPMASTER" (in techinfo.txt)

* Version 1.5:
  - Address mapping added (see ADDRESS MAPPING in manual.txt)
  - Servers on a loopback address are accepted again if they have a mapping
  - A valid "infoReponse" is now rejected if its challenge has timed out
  - The size of the challenge sent with "getinfo" has been made random
  - A timed-out server is now removed as soon as a new server needs a slot
  - Several little changes in the printings to make them more informative
  - A technical documentation was added
  - Compiling dpmaster with MSVC works again

* Version 1.4:
  - Dpmaster now quits if it can't chroot, switch privileges or daemonize
  - Packets coming from a loopback address are now rejected with a warning
  - Listen address option added (-l)
  - Modified Makefile to please BSD make

* Version 1.3.1:
  - SECURITY WARNING: 2 exploitable buffer overflows were fixed
  - Verbose option parsing fixed
  - Paranoid buffer overflow checkings added, in case of future code changes

* Version 1.3:
  - Ability to support any game which uses DP master protocol (ex: QFusion)

* Version 1.2.1:
  - A major bug was fixed (a NULL pointer dereference introduced in v1.2)

* Version 1.2:
  - A major bug was fixed (an infinite loop in HandleGetServers)

* Version 1.1:
  - A lot of optimizations and tweakings
  - Verbose option added (-v)
  - Hash size option added (-H)
  - Daemonization option added on UNIXes (-D)
  - Chrooting and privileges dropping when running as root added on UNIXes
  - MinGW cross-compilation support added

* Version 1.01:
  - A major bug was fixed. Most of the servers weren't sent to the clients

* Version 1.0 :
  - First publicly available version


--
Mathieu Olivier
molivier, at users.sourceforge.net


messages.c
/*
	messages.c

	Message management for dpmaster

	Copyright (C) 2004-2006  Mathieu Olivier

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/


#include "common.h"
#include "messages.h"
#include "servers.h"


// ---------- Constants ---------- //

// Timeouts (in secondes)
#define TIMEOUT_HEARTBEAT		2
#define TIMEOUT_INFORESPONSE	(15 * 60)

// Period of validity for a challenge string (in secondes)
#define TIMEOUT_CHALLENGE 2

// Gamename used for Q3A
#define GAMENAME_Q3A "Quake3Arena"

// Maximum size of a reponse packet
#define MAX_PACKET_SIZE 1400


// Types of messages (with samples):

// Q3: "heartbeat QuakeArena-1\x0A"
// DP: "heartbeat DarkPlaces\x0A"
// QFusion: "heartbeat QFusion\x0A"
#define S2M_HEARTBEAT "heartbeat"

// Q3 & DP & QFusion: "getinfo A_Challenge"
#define M2S_GETINFO "getinfo"

// Q3 & DP & QFusion: "infoResponse\x0A\\pure\\1\\..."
#define S2M_INFORESPONSE "infoResponse\x0A"

// Q3: "getservers 67 empty full"
// DP: "getservers DarkPlaces-Quake 3 empty full"
// DP: "getservers Transfusion 3 empty full"
// QFusion: "getservers qfusion 39 empty full"
#define C2M_GETSERVERS "getservers "

// Q3 & DP & QFusion:
// "getserversResponse\\...(6 bytes)...\\...(6 bytes)...\\EOT\0\0\0"
#define M2C_GETSERVERSREPONSE "getserversResponse"



// ---------- Private functions ---------- //

/*
====================
SearchInfostring

Search an infostring for the value of a key
====================
*/
static char* SearchInfostring (const char* infostring, const char* key)
{
	static char value [256];
	char crt_key [256];
	size_t value_ind, key_ind;
	char c;

	if (*infostring++ != '\\')
		return NULL;

	value_ind = 0;
	for (;;)
	{
		key_ind = 0;

		// Get the key name
		for (;;)
		{
			c = *infostring++;

			if (c == '\0')
				return NULL;
			if (c == '\\' || key_ind == sizeof (crt_key) - 1)
			{
				crt_key[key_ind] = '\0';
				break;
			}

			crt_key[key_ind++] = c;
		}

		// If it's the key we are looking for, save it in "value"
		if (!strcmp (crt_key, key))
		{
			for (;;)
			{
				c = *infostring++;

				if (c == '\0' || c == '\\' || value_ind == sizeof (value) - 1)
				{
					value[value_ind] = '\0';
					return value;
				}

				value[value_ind++] = c;
			}
		}

		// Else, skip the value
		for (;;)
		{
			c = *infostring++;

			if (c == '\0')
				return NULL;
			if (c == '\\')
				break;
		}
	}
}


/*
====================
BuildChallenge

Build a challenge string for a "getinfo" message
====================
*/
static const char* BuildChallenge (void)
{
	static char challenge [CHALLENGE_MAX_LENGTH];
	size_t ind;
	size_t length = CHALLENGE_MIN_LENGTH - 1;  // We start at the minimum size

	// ... then we add a random number of characters
	length += rand () % (CHALLENGE_MAX_LENGTH - CHALLENGE_MIN_LENGTH + 1);

	for (ind = 0; ind < length; ind++)
	{
		char c;
		do
		{
			c = 33 + rand () % (126 - 33 + 1);  // -> c = 33..126
		} while (c == '\\' || c == ';' || c == '"' || c == '%' || c == '/');

		challenge[ind] = c;
	}

	challenge[length] = '\0';
	return challenge;
}


/*
====================
SendGetInfo

Send a "getinfo" message to a server
====================
*/
static void SendGetInfo (server_t* server)
{
	char msg [64] = "\xFF\xFF\xFF\xFF" M2S_GETINFO " ";
	size_t msglen;

	if (!server->challenge_timeout || server->challenge_timeout < crt_time)
	{
		const char* challenge;

		challenge = BuildChallenge ();
		strncpy (server->challenge, challenge, sizeof (server->challenge) - 1);
		server->challenge_timeout = crt_time + TIMEOUT_CHALLENGE;
	}

	msglen = strlen (msg);
	strncpy (msg + msglen, server->challenge, sizeof (msg) - msglen - 1);
	msg[sizeof (msg) - 1] = '\0';
	sendto (sock, msg, strlen (msg), 0,
			(const struct sockaddr*)&server->address,
			sizeof (server->address));

	MsgPrint (MSG_DEBUG, "> %s <--- getinfo with challenge \"%s\"\n",
			  peer_address, server->challenge);
}


/*
====================
HandleGetServers

Parse getservers requests and send the appropriate response
====================
*/
static void HandleGetServers (const char* msg, const struct sockaddr_in* addr)
{
	const char* packetheader = "\xFF\xFF\xFF\xFF" M2C_GETSERVERSREPONSE "\\";
	const size_t headersize = strlen (packetheader);
	char gamename [GAMENAME_LENGTH] = "";
	qbyte packet [MAX_PACKET_SIZE];
	size_t packetind;
	server_t* sv;
	unsigned int protocol;
	unsigned int sv_addr;
	unsigned short sv_port;
	qboolean no_empty;
	qboolean no_full;

	// Check if there's a name before the protocol number
	// In this case, the message comes from a DarkPlaces-compatible client
	protocol = atoi (msg);
	if (!protocol)
	{
		char *space;

		strncpy (gamename, msg, sizeof (gamename) - 1);
		gamename[sizeof (gamename) - 1] = '\0';
		space = strchr (gamename, ' ');
		if (space)
			*space = '\0';
		msg += strlen (gamename) + 1;

		protocol = atoi (msg);
	}
	// Else, it comes from a Quake III Arena client
	else
	{
		strncpy (gamename, GAMENAME_Q3A, sizeof (gamename) - 1);
		gamename[sizeof (gamename) - 1] = '\0';
	}

	MsgPrint (MSG_NORMAL, "> %s ---> getservers (%s)\n", peer_address,
			  gamename);

	no_empty = (strstr (msg, "empty") == NULL);
	no_full = (strstr (msg, "full") == NULL);

	// Initialize the packet contents with the header
	packetind = headersize;
	memcpy(packet, packetheader, headersize);

	// Add every relevant server
	for (sv = Sv_GetFirst (); /* see below */;  sv = Sv_GetNext ())
	{
		// If we're done, or if the packet is full, send the packet
		if (sv == NULL || packetind > sizeof (packet) - (7 + 6))
		{
			// End Of Transmission
			packet[packetind    ] = 'E';
			packet[packetind + 1] = 'O';
			packet[packetind + 2] = 'T';
			packet[packetind + 3] = '\0';
			packet[packetind + 4] = '\0';
			packet[packetind + 5] = '\0';
			packetind += 6;

			// Send the packet to the client
			sendto (sock, packet, packetind, 0, (const struct sockaddr*)addr,
					sizeof (*addr));

			MsgPrint (MSG_DEBUG, "> %s <--- getserversResponse (%u servers)\n",
						peer_address, (packetind - headersize - 1) / 7 - 1);

			// If we're done
			if (sv == NULL)
				return;
			
			// Reset the packet index (no need to change the header)
			packetind = headersize;
		}

		sv_addr = ntohl (sv->address.sin_addr.s_addr);
		sv_port = ntohs (sv->address.sin_port);

		// Extra debugging info
		if (max_msg_level >= MSG_DEBUG)
		{
			MsgPrint (MSG_DEBUG,
					  "Comparing server: IP:\"%u.%u.%u.%u:%hu\", p:%u, c:%hu, g:\"%s\"\n",
					  sv_addr >> 24, (sv_addr >> 16) & 0xFF,
					  (sv_addr >>  8) & 0xFF, sv_addr & 0xFF,
					  sv_port, sv->protocol, sv->nbclients, sv->gamename);

			if (sv->protocol != protocol)
				MsgPrint (MSG_DEBUG,
						  "Reject: protocol %u != requested %u\n",
						  sv->protocol, protocol);
			if (sv->nbclients == 0 && no_empty)
				MsgPrint (MSG_DEBUG,
						  "Reject: nbclients is %hu/%hu && no_empty\n",
						  sv->nbclients, sv->maxclients);
			if (sv->nbclients == sv->maxclients && no_full)
				MsgPrint (MSG_DEBUG,
						  "Reject: nbclients is %hu/%hu && no_full\n",
						  sv->nbclients, sv->maxclients);
			if (gamename[0] && strcmp (gamename, sv->gamename))
				MsgPrint (MSG_DEBUG,
						  "Reject: gamename \"%s\" != requested \"%s\"\n",
						  sv->gamename, gamename);
		}

		// Check protocol, options, and gamename
		if (sv->protocol != protocol ||
			(sv->nbclients == 0 && no_empty) ||
			(sv->nbclients == sv->maxclients && no_full) ||
			(gamename[0] && strcmp (gamename, sv->gamename)))
		{

			// Skip it
			continue;
		}

		// Use the address mapping associated with the server, if any
		if (sv->addrmap != NULL)
		{
			const addrmap_t* addrmap = sv->addrmap;

			sv_addr = ntohl (addrmap->to.sin_addr.s_addr);
			if (addrmap->to.sin_port != 0)
				sv_port = ntohs (addrmap->to.sin_port);

			MsgPrint (MSG_DEBUG,
					  "Server address mapped to %u.%u.%u.%u:%hu\n",
					  sv_addr >> 24, (sv_addr >> 16) & 0xFF,
					  (sv_addr >>  8) & 0xFF, sv_addr & 0xFF,
					  sv_port);
		}

		// IP address
		packet[packetind    ] =  sv_addr >> 24;
		packet[packetind + 1] = (sv_addr >> 16) & 0xFF;
		packet[packetind + 2] = (sv_addr >>  8) & 0xFF;
		packet[packetind + 3] =  sv_addr        & 0xFF;

		// Port
		packet[packetind + 4] = sv_port >> 8;
		packet[packetind + 5] = sv_port & 0xFF;

		// Trailing '\'
		packet[packetind + 6] = '\\';

		MsgPrint (MSG_DEBUG, "  - Sending server %u.%u.%u.%u:%hu\n",
				  packet[packetind    ], packet[packetind + 1],
				  packet[packetind + 2], packet[packetind + 3],
				  sv_port);

		packetind += 7;
	}
}


/*
====================
HandleInfoResponse

Parse infoResponse messages
====================
*/
static void HandleInfoResponse (server_t* server, const char* msg)
{
	char* value;
	unsigned int new_protocol = 0, new_maxclients = 0;

	MsgPrint (MSG_DEBUG, "> %s ---> infoResponse\n", peer_address);

	// Check the challenge
	if (!server->challenge_timeout || server->challenge_timeout < crt_time)
	{
		MsgPrint (MSG_WARNING,
				  "> WARNING: infoResponse with obsolete challenge from %s\n",
				  peer_address);
		return;
	}
	value = SearchInfostring (msg, "challenge");
	if (!value || strcmp (value, server->challenge))
	{
		MsgPrint (MSG_ERROR, "> ERROR: invalid challenge from %s (%s)\n",
				  peer_address, value);
		return;
	}

	// Check and save the values of "protocol" and "maxclients"
	value = SearchInfostring (msg, "protocol");
	if (value)
		new_protocol = atoi (value);
	value = SearchInfostring (msg, "sv_maxclients");
	if (value)
		new_maxclients = atoi (value);
	if (!new_protocol || !new_maxclients)
	{
		MsgPrint (MSG_ERROR,
				  "> ERROR: invalid infoResponse from %s (protocol: %d, maxclients: %d)\n",
				  peer_address, new_protocol, new_maxclients);
		return;
	}
	server->protocol = new_protocol;
	server->maxclients = new_maxclients;

	// Save some other useful values
	value = SearchInfostring (msg, "clients");
	if (value)
		server->nbclients = atoi (value);
	value = SearchInfostring (msg, "gamename");

	// Q3A doesn't send a gamename, so we add it manually
	if (value == NULL)
		value = GAMENAME_Q3A;

	// If the gamename has changed
	if (strcmp (server->gamename, value))
	{
		MsgPrint (MSG_NORMAL,
				  "> Server %s updated its gamename: \"%s\" -> \"%s\"\n",
				  peer_address, server->gamename, value);

		strncpy (server->gamename, value, sizeof (server->gamename) - 1);
	}

	// Set a new timeout
	server->timeout = crt_time + TIMEOUT_INFORESPONSE;
}


// ---------- Public functions ---------- //

/*
====================
HandleMessage

Parse a packet to figure out what to do with it
====================
*/
void HandleMessage (const char* msg, size_t length,
					const struct sockaddr_in* address)
{
	server_t* server;

	// If it's an heartbeat
	if (!strncmp (S2M_HEARTBEAT, msg, strlen (S2M_HEARTBEAT)))
	{
		char gameId [64];

		// Extract the game id
		sscanf (msg + strlen (S2M_HEARTBEAT) + 1, "%63s", gameId);
		MsgPrint (MSG_DEBUG, "> %s ---> heartbeat (%s)\n",
				  peer_address, gameId);

		// Get the server in the list (add it to the list if necessary)
		server = Sv_GetByAddr (address, true);
		if (server == NULL)
			return;

		server->active = true;

		// If we haven't yet received any infoResponse from this server,
		// we let it some more time to contact us. After that, only
		// infoResponse messages can update the timeout value.
		if (!server->maxclients)
			server->timeout = crt_time + TIMEOUT_HEARTBEAT;

		// Ask for some infos
		SendGetInfo (server);
	}

	// If it's an infoResponse message
	else if (!strncmp (S2M_INFORESPONSE, msg, strlen (S2M_INFORESPONSE)))
	{
		server = Sv_GetByAddr (address, false);
		if (server == NULL)
			return;

		HandleInfoResponse (server, msg + strlen (S2M_INFORESPONSE));
	}

	// If it's a getservers request
	else if (!strncmp (C2M_GETSERVERS, msg, strlen (C2M_GETSERVERS)))
	{
		HandleGetServers (msg + strlen (C2M_GETSERVERS), address);
	}
}
